### DISCUSSION TITLE
ComboBox should let user use search query as input

### DISCUSSION THREAD

**Discussion by zhuhaow** [tailwindlabs/headlessui#1228] [Ideas]
One useful scenario for ComboBox is to use for search queries, where user will see query suggestions, but user doesn't have to choose from the suggestions.

However, my test suggests ComboBox requires the user to select from the dropdown list otherwise the `onChange` won't be triggered when the user presses `enter`. Maybe there is something I'm missing or this case is not supported yet.

Any help would be appreciated.

**Comment by lordvcs** [comment]
I am interested in this too. If this component is truly meant as an autocomplete component the user should be able to enter query that may not show up in the autocomplete suggestions dropdown.

**Comment by RobinMalfait** [comment]
Hey! Thank you for your suggestion!
Much appreciated! 🙏

I'm not 100% sure what you mean. But is your goal to allow to let's say create a new record (based on the query) if it doesn't exist in a certain list?

Is this what you mean? https://codesandbox.io/s/condescending-grass-konu6w?file=/src/App.js

**Reply by zhuhaow** [reply]
Thanks for the quick reply.

Sorry I didn't put this clearly. 

The current behavior of ComboBox is, if user input something that doesn't match a record and press enter, it will select the previously selected item. This behavior will not work for cases where we want to use ComboBox as a search input with query suggestions such as Google search.

**Reply by lordvcs** [reply]
@RobinMalfait A good example would be the google search, the google search input shows all the suggested items in the dropdown, u can either choose a suggested dropdown item or use ur own input by pressing enter.

**Comment by kristiandupont** [comment]
I am interested in this as well. Basically a high-fidelity version of `<input type="search" />`. It looks like react-aria has something, so I will try that out: https://react-spectrum.adobe.com/react-aria/useSearchField.html

**Comment by ragulka** [comment]
I can get behind this, as well. I would be great if we can use ComboBox to simply show suggestions. The user would be free to use them or ignore them altogether. I understand there are use-cases for both scenarios (ie, sometimes you want to force the user to make a selection).

**Reply by RobinMalfait** [reply]
This is already possible, unless you want something else? 

- https://headlessui.dev/react/combobox#allowing-custom-values
- https://headlessui.dev/vue/combobox#allowing-custom-values

**Reply by ragulka** [reply]
How about when I don't want to show a "create" option? I'd like to silently allow a custom value.

**Reply by ragulka** [reply]
See https://www.w3.org/TR/wai-aria-practices-1.1/examples/combobox/aria1.1pattern/listbox-combo.html Example 1: List Autocomplete with Manual Selection. No requirement to select anything, not resetting the input value when user does not make a selection.

**Reply by mathijsfr** [reply]
Link doesn't work anymore

**Reply by ragulka** [reply]
Looks like W3 have update these examples. Please see the following ones:
* https://www.w3.org/WAI/ARIA/apg/example-index/combobox/combobox-autocomplete-list.html
* https://www.w3.org/WAI/ARIA/apg/example-index/combobox/combobox-autocomplete-both.html

**Reply by stephent** [reply]
@RobinMalfait is it the case that https://headlessui.dev/react/combobox#allowing-custom-values should already do what is demonstrated by https://www.w3.org/WAI/ARIA/apg/example-index/combobox/combobox-autocomplete-list.html ? (I'm struggling to get the [example](https://headlessui.dev/react/combobox#allowing-custom-values) to work.)

**Reply by ragulka** [reply]
@RobinMalfait I'd like to circle back to this. ARIA guidelines describe the different implementations of a combobox:

> In some implementations, the popup presents allowed values, while in other implementations, **the popup presents suggested values**, and users **may either select one of the suggestions or type a value**

Admittedly, https://headlessui.dev/vue/combobox#allowing-custom-values _does_ allow using a custom value, but the implementation is rather restrictive. It does not allow the user to ignore the custom option. The user _must_ make a choice to either select the custom option or discard their input. This is not how the ARIA combobox examples are implemented.

See the [Editable Combobox With List Autocomplete Example](https://www.w3.org/WAI/ARIA/apg/example-index/combobox/combobox-autocomplete-list.html#ex_label):
1. Type a custom value in the field, such as `Something`
2. Note that there are no suggestions displayed
3. Move focus away from the field
4. Note that the combobox still shows `Something` as the value (it does not reset to an empty string or predefined suggestion)

 In Headless UI combobox, this is not possible. We always _need_ to show a custom option with the custom value the user typed in. And, if the user does not explicitly select the custom option, the combobox value will reset to what it was previously.

**Comment by haykkh** [comment]
I implemented this by having an invisible "insert custom value" `<ComboboxOption>` and being picky about what value I emit up

```vue
<script setup lang="ts">
const query = ref('')

// this can't return null because we need headlessui to pick always see it
const customValue = computed(() => (query.value === '' ? {} : { id: null, name: query.value }))

const updateValue = (value) => {
  // if the user has typed in a custom value always prioritize that, otherwise take what headlessui gives you
  emit('update:modelValue', customValue.value?.name ? customValue.value : value)

  // reset query.value for future interactions
  query.value = ''
}
</script>

<template>
  <Combobox :model-value="modelValue" @update:model-value="updateValue">
  ...

  <ComboboxOptions>
    <ComboboxOption :value="customValue" class="h-0 w-0 overflow-hidden"></ComboboxOption>
    ...
  </ComboboxOptions>

  ...
</template>
```

**Comment by designbyadrian** [comment]
I agree with @ragulka 

To make Combobox have universal use, it should at least feature:

* free text input that does not clear on blur
* a clear entry button like input[type=search]
* an event that fires when user hits enter in the input box

Much like react-aria as demonstrated here:

https://codesandbox.io/s/hardcore-moon-xzc4r?file=/src/Popover.tsx

**Comment by mdashfaque-mamdu-au17** [comment]
@RobinMalfait is there any workaround for this one. any ideas on how we can achieve this using headlessui

**Comment by RobinMalfait** [comment]
Hey! Here is an example on how you can achieve something like this: https://headlessui.com/react/combobox#allowing-custom-values

**Reply by ragulka** [reply]
@RobinMalfait that example is _nothing_ like me and others are advocating for here. Please see my [original comment from January 10th](https://github.com/tailwindlabs/headlessui/discussions/1228#discussioncomment-4642846) with the WAI ARIA example.

**Reply by RobinMalfait** [reply]
I see what you mean, sorry I wasn't clear.

What I meant was that we currently don't support _exactly_ what you want, but there is a workaround that gets you close to the functionality where you can allow any value instead of choosing from a pre-defined list. It's true that this will require you to add a "create <query>" like option, and if you don't want that behaviour then that's something we unfortunately don't support right now.

I can't promise an ETA for when and even if we will add the other "mode" where you don't have to choose from a predefined list of options in the Combobox.

**Reply by ragulka** [reply]
I understand this hasn't been on the horizon. However, would you consider a PR that at least showcases this mode in ComboBox - even if at a very basic level? I can hack together a working solution, but it will probably not be at the level of abstraction or sophistication that I see in this library. 

Basically, if there is a community effort to add this "mode", would you be able to guide or flesh out the details that the community may not be aware of?

**Reply by pixelmultiplo** [reply]
Whoa this is a pretty big lack. Material UI has the freeSolo attribute to solve it in 2 seconds.

**Comment by pixelmultiplo** [comment]
So the Combobox cannot be used for an Autocomplete. What should i use?

**Reply by TuanPhamHP** [reply]
a bit tricky but you can custom the @after-leave (in case you using Vue) event.

```
<TransitionRoot
         ...some-config-here
	@after-leave="handleAfterLeave"
>
```

and this is the custom function:

```
handleAfterLeave() {
	  if (this.canBeCustomValue && this.query) {
		  const rs = {
			  id: this.query,
			  name: this.query,
		  };
		  this.customData.push(rs);
		  this.selectedData = rs.id;
	  }
	  this.query = '';
	},
```
so, i'm just create new props 'canBeCustomValue' to enable this behavior, when you 'leave' (blur...) the <Autocomplete> will create new option and select this as well. And yes, its not a good solution like 'MUI', hope that headlessui team can publish an update about things like this.

**Comment by wabawden** [comment]
If anyone is still interested in this: my solution was to place a hidden ComboBoxOption before the rest of the options, with a value that tracks the input value. For keyboard navigation I feel this demonstrates expected behaviour for enter and arrow keys (e.g. pressing down will now select the first suggestion), however one minor issue is that pressing 'home' will select the hidden value when you would probably expect it to select the first suggestion (I suppose you could add a keydown handler for this if you really wanted).

`

            <ComboboxOptions className="absolute z-10 mt-1 max-h-60 w-full overflow-auto rounded-md border bg-white shadow-lg">
              <ComboboxOption
                value={{ displayName: airportName } as google.maps.places.Place}
                className="hidden"
              >
                {airportName}
              </ComboboxOption>
              {nearbyAirports.length > 0 && (
                <>
                  {nearbyAirports.map((a) => (
                    <ComboboxOption
                      key={a.id}
                      value={a}
                      className={({ active }) =>
                        `cursor-pointer px-4 py-2 ${
                          active ? "bg-blue-100" : ""
                        }`
                      }
                    >
                      {a.displayName}
                    </ComboboxOption>
                  ))}
                </>
              )}
            </ComboboxOptions>
`
